tests/test_parser.py

```diff
@@ -22,11 +22,38 @@
 )
 from prometheus_client.parser import (
     text_string_to_metric_families,
+    _replace_escaping, _replace_help_escaping
 )
 
 
 class TestParse(unittest.TestCase):
 
+    def test_replace_escaping(self):
+        self.assertEqual(_replace_escaping('foo'), 'foo')
+        self.assertEqual(_replace_escaping('\\foo'), '\\foo')
+        self.assertEqual(_replace_escaping('\\\\foo'), '\\foo')
+        self.assertEqual(_replace_escaping('foo\\'), 'foo\\')
+        self.assertEqual(_replace_escaping('foo\\\\'), 'foo\\')
+        self.assertEqual(_replace_escaping('\\n'), '\n')
+        self.assertEqual(_replace_escaping('\\\\n'), '\\n')
+        self.assertEqual(_replace_escaping('\\\\\\n'), '\\\n')
+        self.assertEqual(_replace_escaping('\\"'), '"')
+        self.assertEqual(_replace_escaping('\\\\"'), '\\"')
```

Comment:
Ok, what would you like to do about that invalid input: fix the parser to raise an error or just not test for it?

The initial version had `r`s, but it doesn't work for strings that end with backslashes (`r'\'`), so then you have to put some strings with `r` and some without, and it felt like it added a bit to the confusion.

Reply:
I'd mostly look to test this a level up with whole lines.

Reply:
Good call. I've moved it to work with input lines instead and also found bugs in escaping handling in _parse_labels, so I proceeded to rewrite all the parser on top of `re`.

Reply:
There were just some changes contributed to improve performance, and I'd rather not change all that again with an RE implementation rather than the explicit FSM that's in use. If you can fix it under the current setup so thsat at least all valid input is correctly handled, that's sufficient.

Reply:
I'm not sure I see how this PR is different to 282. Before 282 there was an explicit FSM implementation, and it was replaced by an explicit algorithmic approach that's easy to get wrong, which is exactly what happened. This PR suggests to replace the algorithmic approach with regexps, because they offer a concise and easy-to-understand way to work with sequences of characters that also happens to be faster in some cases than pure-python implementations. Could you elaborate what's wrong with it?

Reply:
You're changing from a standard FSM approach for a parser, to an unusual mix of FSM and regex. This is confusing and will be harder to maintain. Please restrict your fix to only what it needs to change.

Reply:
I'd like to restate that the departure from a standard FSM approach to a somewhat confusing one has already happened in #282.

I don't think I agree that regexes are unusual in the parsing business, they have been around at least since lex & yacc duo which were introduced in the 1970s.

But OK, the decision is up to you.